pkg/operator/certupdater/cert_updater.go

```diff
@@ -0,0 +1,212 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package certupdater contains an implementation of `tls.GetCertificate`.
+package certupdater
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"io/fs"
+	"os"
+	"sync"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/cert"
+)
+
+const TLSCAKey = "ca.crt"
+
+type CertUpdater struct {
+	sync.RWMutex
+
+	source      CertSource
+	currentCert *tls.Certificate
+	currentCA   *x509.Certificate
+
+	pollingInterval time.Duration
+}
+
+type Option func(*CertUpdater)
+
+func WithPolling(d time.Duration) Option {
+	return func(cu *CertUpdater) {
+		cu.pollingInterval = d
+	}
+}
+
+func New(source CertSource, opts ...Option) *CertUpdater {
+	cu := &CertUpdater{
+		source: source,
+	}
+	return cu
+}
+
+// Start begins polling to periodically update certificates from the available sources.
+func (cu *CertUpdater) Start(ctx context.Context) error {
+	go func() {
+		done := cu.pollingInterval == 0
+		for !done {
+			cu.poll(ctx)
+
+			select {
+			case <-time.After(cu.pollingInterval):
+				// Noop
+			case <-ctx.Done():
+				done = true
+			}
+		}
+	}()
+	return nil
+}
+
+func (cu *CertUpdater) poll(ctx context.Context) error {
+	if cu.source == nil {
```

Comment:
Assuming you're referring to `Start()`?

As a general rule, I don't like spawning long-running processes in a constructor. I feel like that's unexpected behavior and may interfere with some testing scenarios.

Reply:
I was thinking to `New` and make it return `(*CertUpdater, error)`. This feels like an initialization error to me so to keep checking this in a loop is inefficient. Afterall, the source cannot change after the constructor. Thoughts?

Reply:
Oh! You're talking about `if cu.source == nil`.

I think this is necessary because it's possible (although inadvisable) to create a CertUpdater literal. `cu := CertUpdater{}` which would not have a source on it. But I should be able to make the type unexported.